packages/compiler/src/core/diagnostics.ts

```diff
@@ -111,6 +111,40 @@ export function getSourceLocation(
   }
 }
 
+/**
+ * @internal
+ */
+export function getDiagnosticNodeStack(
+  target: DiagnosticTarget | typeof NoTarget | undefined,
+): Node[] {
+  if (typeof target !== "object" || !("entityKind" in target) || !("templateMapper" in target)) {
+    return [];
+  }
+
+  const result = [];
+  let current = target.templateMapper;
+  while (current) {
+    result.push(current.source.node);
+    current = current.source.mapper;
+  }
+  return result;
+}
+
+/**
+ * @internal
+ */
+export function getSourceLocationStack(
```

Comment:
Wondering if we shouldn't export this one at all and let things call `getDiagnosticNodeStack` and rename to `getDiagnosticTraceNodes` or something like that